Increase leader-election nominal concurrency shares from 10 to 40

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation


Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
Actually, kind of hard to classify. This is a tweak to an existing feature, based on results from experience and non-CI testing.

#### What this PR does / why we need it:
This PR increases the default setting for the nominal concurrency shares of the `leader-election` priority level in the API Priority and Fairness feature. Experience shows that in some stressful situations APF gives too little concurrency to the leader election requests, which causes the corresponding controller(s) to go out of business (and, due to the stress, a replacement has trouble getting up and elected). This PR changes the default setting of that level's nominal concurrency shares from 10 to 40.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The default nominal concurrency shares of the "leader-election" priority level in the API Priority and Fairness feature has been changed from 10 to 40.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Increase leader-election nominal concurrency shares from 10 to 40

Because some testing shows leader election being starved.

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer commented:
/cc @linxiulei

fedebongio commented:
/triage accepted

MikeSpreitzer commented:
@linxiulei compared two runs with and without tweaking the nominal concurrency shares of the leader-election priority level. I have graphed the results, see https://docs.google.com/document/d/1DEexuOLOTmtNDAcAgnSOlVGsDXcq9f5_f_V77qYB3ME and https://docs.google.com/document/d/1Bz4t6A_H1z-jmsOL-Fa05EW9e0DL1uvtb9JzqwXw3bY

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:
```diff
@@ -208,7 +208,7 @@ var (
 			Type: flowcontrol.PriorityLevelEnablementLimited,
 			Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
-				NominalConcurrencyShares: ptr.To(int32(10)),
+				NominalConcurrencyShares: ptr.To(int32(40)),
```
Hmm - with this change we're going up from default total of 245 to 275 without changing the lendable percent.
So technically, for usecases that don't need it, we effectively decrease the capacity by 10%.

Shouldn't we also adjust the LendablePercent here?

MikeSpreitzer replied:
Interesting suggestion. I have added a commit that increases the lendable percent to 75, thus keeping the unlendable shares at 10 like before this PR. There is a small difference due to the increase in total nominal shares, but I suspect that will be unimportant.

wojtek-t replied:
Thank you!

> There is a small difference due to the increase in total nominal shares, but I suspect that will be unimportant.

I looked into those numbers - there are changes, for typical values that we use in GKE, I see differences of actually available seats <15%. This isn't negligible, but I also don't know how to assess the risk of such a change.

With borrowing, in overloaded cluster I think that the risk will be mitigated, but I think we need ability to disable this change - so I would like to hide this change behind a feature-gate (though it can be enabled by default).

MikeSpreitzer replied:
Remember that we have two other config changes being considered now:
1. Segregating events (introduced in #128974);
2. rejiggering to stop working around the lack of borrowing (https://github.com/kubernetes/kubernetes/issues/121982#issuecomment-1933534198)

MikeSpreitzer committed:
Increase leader-election's lendable percent

.. so that it takes no larger bite from system capacity than before
(modulo small difference due to total nominal shares).
